src/sink/kafka.rs

```diff
@@ -0,0 +1,214 @@
+pub struct KafkaConfig {
+    /// Canonical name for the given Kafka sink.
+    pub config_path: Option<String>,
+    /// Kafka topic to publish to.
+    pub topic_name: Option<String>,
+    /// Kafka brokers. This is a comma-separated list of host or host:port.
+    pub brokers: Option<String>,
+    /// Underlying librdkafka configuration.
+    pub rdkafka_config: Option<HashMap<String, String>>,
+    /// How often (seconds) the in-flight messages are checked for delivery.
+    /// Default = 1 second
+    pub flush_interval: u64,
+}
+
+impl Default for KafkaConfig {
+    fn default() -> KafkaConfig {
+        KafkaConfig {
+            config_path: None,
+            topic_name: None,
+            brokers: None,
+            rdkafka_config: None,
+            flush_interval: 1,
+        }
+    }
+}
+
+/// Kafka sink internal state.
+pub struct Kafka {
+    /// Name of the stream we are publishing to.
+    topic_name: String,
+    /// A message producers.
+    producer: FutureProducer<EmptyContext>,
+    // In-flight messages.
+    messages: Vec<DeliveryFuture>,
+    /// How often (seconds) the in-flight messages are checked for delivery.
+    flush_interval: u64,
+}
+
+impl Sink<KafkaConfig> for Kafka {
+    fn init(config: KafkaConfig) -> Self {
+        if config.topic_name.is_none() {
+            panic!("No Kafka topic name provided!");
+        }
+        if config.brokers.is_none() {
+            panic!("No Kafka brokers provided!")
+        }
+
+        let mut producer_config = ClientConfig::new();
+        if let Some(ref map) = config.rdkafka_config {
+            for (key, value) in map.iter() {
+                producer_config.set(key, value);
+            }
+        }
+        producer_config.set("bootstrap.servers", &config.brokers.unwrap()[..]);
+
+        Kafka {
+            topic_name: config.topic_name.unwrap(),
+            producer: producer_config.create::<FutureProducer<_>>().unwrap(),
+            messages: Vec::new(),
+            flush_interval: config.flush_interval,
+        }
+    }
+
+    fn valve_state(&self) -> Valve {
+        // We never close up shop.
+        Valve::Open
+    }
+
+    fn deliver(&mut self, _: Arc<Option<Telemetry>>) -> () {
+        // Discard point
+    }
+
+    fn deliver_line(&mut self, _: Arc<Option<LogLine>>) -> () {
+        // Discard line
+    }
+
+    /// Fire off the given event to librdkafka. That library handles buffering and
+    /// batching internally.
+    fn deliver_raw(
+        &mut self,
+        order_by: u64,
+        _encoding: metric::Encoding,
+        bytes: Vec<u8>,
+    ) {
+        let key = format!("{:X}", order_by);
+        let future = self.try_payload(bytes.as_slice(), key.as_bytes());
+        self.messages.push(future);
+    }
+
+    fn flush(&mut self) {
+        let retry_payload_and_keys: Vec<Option<OwnedMessage>> = self.messages
+            .iter_mut()
+            .map(|future| {
+                let result = future.wait();
```

Comment:
I wonder if by waiting here we aren't shooting ourselves in the foot some. I _think_ what's going to happen is that rdkafka's IO loop will be blocked -- correct me if I'm wrong -- and we're going to see much worse throughput as a result. _If_ that is the case maybe what we ought to do every flush is troll through the futures in self.messages, poll them, if they're cleared out then remove them from self.messages? 

Would imply a need for a valve, still.

Reply:
I'm happy to change this logic around if it ends up having crappy performance. First, I want to see (via a benchmark) how well it does perform in this simpler state though. 

As for the IO loop blocking, I don't think thats's the case. The future producer wraps a threaded producer and starts a thread that is constantly polling for the future's to complete:

Here's where the FutureProducer creates and wraps a ThreadedProducer:
https://github.com/fede1024/rust-rdkafka/blob/master/src/producer/future_producer.rs#L84

Here's where the ThreadedProducer is created and the thread started.
https://github.com/fede1024/rust-rdkafka/blob/master/src/producer/base_producer.rs#L244

The start loop is a few lines below there.

Reply:
Cool. That works back my concern then.